launcher/launcher.go

```diff
@@ -193,11 +208,34 @@ func (l *launcher) stopBeater() {
 	l.log.Infof("Launcher shut %s down gracefully", l.name)
 }
 
-// waitForUpdates is the function that keeps Launcher runLoop busy
+// Returns an error indicating if the beater was stopped gracefully or not
+func (l *launcher) stopBeaterWithTimeout(duration time.Duration) error {
+	l.log.Infof("Launcher is shutting %s down gracefully", l.name)
+	l.beater.Stop()
+
+	wgCh := make(chan struct{})
+
+	go func() {
+		// By waiting to the wait group, we make sure that the old beater has really stopped
+		l.wg.Wait()
+		close(wgCh)
+	}()
+
+	select {
+	case <-time.After(duration):
+		l.log.Infof("Grace period for %s ended", l.name)
+		return ErrorTimeoutExit
+	case <-wgCh:
+		l.log.Infof("Launcher shut %s down gracefully", l.name)
+		return ErrorGracefulExit
+	}
+}
+
+// waitForUpdates is the function that keeps Launcher runLoop busy.
 // It will finish for one of following reasons:
-// 1. The Stop function got called
-// 2. The beater run has returned
-// 3. A config update received
+//  1. The Stop function got called 	(nil, nil)
+//  2. The beater run has returned 		(nil, err)
+//  3. A config update received 		(val, nil)
 func (l *launcher) waitForUpdates() (*config.C, error) {
 	select {
 	case err := <-l.beaterErr:
```

Comment:
@orouz
wdyt about indicating that a stop function has been called explicitly?
```suggestion
	case err, ok := <-l.beaterErr:
	 if !ok {
	     return nil, fmt.errorf("close channel/stop function has been called")
	 }
```